taichi/ui/backends/vulkan/renderable.cpp

```diff
@@ -0,0 +1,301 @@
+
+  if (app_context_->config.ti_arch == Arch::cuda) {
+    auto [vb_mem, vb_offset, vb_size] =
+        app_context_->device().get_vkmemory_offset_size(vertex_buffer_);
+
+    auto [ib_mem, ib_offset, ib_size] =
+        app_context_->device().get_vkmemory_offset_size(index_buffer_);
+
+    auto block_size = VulkanDevice::kMemoryBlockSize;
+
+    vertex_buffer_device_ptr_ =
+        (Vertex *)get_memory_pointer(vb_mem, block_size, vb_offset, vb_size,
+                                     app_context_->device().vk_device());
+    index_buffer_device_ptr_ =
+        (int *)get_memory_pointer(ib_mem, block_size, ib_offset, ib_size,
+                                  app_context_->device().vk_device());
+  }
+}
+
+void Renderable::update_data(const RenderableInfo &info) {
+  int num_vertices = info.vertices.shape[0];
+  int num_indices;
+  if (info.indices.valid) {
+    num_indices = info.indices.shape[0];
+    if (info.indices.dtype != PrimitiveType::i32 &&
+        info.indices.dtype != PrimitiveType::u32) {
+      throw std::runtime_error("dtype needs to be 32-bit ints for indices");
+    }
+  } else {
+    num_indices = num_vertices;
+  }
+  if (num_vertices > config_.vertices_count ||
+      num_indices > config_.indices_count) {
+    free_buffers();
+    config_.vertices_count = num_vertices;
+    config_.indices_count = num_indices;
+    init_buffers();
+  }
+
+  if (info.vertices.dtype != PrimitiveType::f32) {
+    throw std::runtime_error("dtype needs to be f32 for vertices");
+  }
+
+  int num_components = info.vertices.matrix_rows;
+
+  if (info.vertices.field_source == FieldSource::TaichiCuda) {
+    update_renderables_vertices_cuda(vertex_buffer_device_ptr_,
+                                     (float *)info.vertices.data, num_vertices,
+                                     num_components, offsetof(Vertex, pos));
+
+    if (info.per_vertex_color.valid) {
+      if (info.per_vertex_color.shape[0] != num_vertices) {
+        throw std::runtime_error(
+            "shape of per_vertex_color should be the same as vertices");
+      }
+      update_renderables_vertices_cuda(
+          vertex_buffer_device_ptr_, (float *)info.per_vertex_color.data,
+          num_vertices, 3, offsetof(Vertex, color));
+    }
+
+    if (info.normals.valid) {
+      if (info.normals.shape[0] != num_vertices) {
+        throw std::runtime_error(
+            "shape of normals should be the same as vertices");
+      }
+      update_renderables_vertices_cuda(vertex_buffer_device_ptr_,
+                                       (float *)info.normals.data, num_vertices,
+                                       3, offsetof(Vertex, normal));
+    }
+
+    if (info.indices.valid) {
+      indexed_ = true;
+      update_renderables_indices_cuda(index_buffer_device_ptr_,
+                                      (int *)info.indices.data, num_indices);
+    } else {
+      indexed_ = false;
+    }
+
+  } else if (info.vertices.field_source == FieldSource::TaichiX64) {
+    {
```

Comment:
looks like a good place to let in numpy arrays & not supported backends

Reply:
Yep.

Although I intended `FieldSource::TaichiX64` to represent fields directly from a `ti.field` from a x64 backend. I will add a TODO for a `FieldSource::Numpy` here.

I suspected that the data from a x64 field could be different from those dumped to an np array. There're a few things at play: power-of-2-fields, packed mode, and multiple level dense layouts... I haven't given much thought into how to organise support for all of this.